### PR TITLE
Team mode

### PR DESCRIPTION
**Pull request by FloEdelmann**
I started working on #2189.

Now I'm stuck; I don't know where to implement the actual OSM element filter logic. But I wanted to share my (early) progress anyway. I appreciate any feedback and help on where/how to continue. Also, taking my commits and adding new code yourself is welcome, too!

### CODE REVIEW

**Comment by FloEdelmann** [comment]
### Main menu

There is now a "Team Mode" button. When tapping that, the main menu closes and the team mode dialog opens.



### Team Mode dialog

#### Choosing team size

When entering an out-of-bounds number (I allowed team sizes of 2 to  12, quite arbitrarily), a hint is shown.

 

#### Choosing an avatar

When entering a valid number, that many avatars are displayed to choose from, alongside an explanation that every team member has to choose one on their own phone. A different set of avatars is shown for each number, and on every day ([logic here](https://github.com/FloEdelmann/StreetComplete/blob/d868c4b0002321bea1e7eb44683f74c9816d6fdf/app/src/main/java/de/westnordost/streetcomplete/controls/TeamModeDialog.kt#L110-L116)). I used these SVGs: https://www.svgrepo.com/collection/animals-6/

 

**Issue:** The last row of avatars is always cut off. I couldn't get the [`RecyclerView`](https://github.com/FloEdelmann/StreetComplete/blob/d868c4b0002321bea1e7eb44683f74c9816d6fdf/app/src/main/res/layout/dialog_team_mode.xml#L70-L81) to scroll properly without being cut off.

#### Confirming

When one avatar is chosen, it's highlighted so that all team members can compare their choice, and an "Ok" button is shown, similar to the quest layout.

 

**Issue:** The button overlaps the last avatar icon. When wrapping the `ConstraintLayout` in a `RelativeLayout` with no padding (so that I can put the button there, so that it is pushed more towards the dialog edges), it completely messes up the `RecyclerView` height and cuts off the number input on the right side.

**Comment by westnordost** [comment]
Wow, really nice! The wording is on point, I think! Thank you for the screenshots!

A few early comments:
- a floating action button in a dialog is very uncommon, it would be more consistent to use the button bar
- I am not sure if I like the "animal avatars". The idea to not let users choose "a number from 1 to number of people" is good and fair, but introducing the concept of avatars may add some confusion as avatars come with certain expectations tied to them. It may be more clear to let the user **choose a color**: red, blue, green, yellow, black, orange, violet etc. 

> Now I'm stuck; I don't know where to implement the actual OSM element filter logic
The logical place would be `VisibleQuestSource`. This class would have another new dependency called `TeamModeQuestFilter` or something like this with one public method like `isVisible(quest: Quest): Boolean`. That class would just always return true if team mode is not on, and if team mode is on, return `quest.id % teamSize == teamIndex`. `TeamModeQuestFilter` would probably save the team mode in the shared preferences.
Then, that class should maybe have another two methods: 1. setting the team mode (on and off) and 2. checking if the team mode is on. Finally, it should be possible to register a listener onto that class so that other parts of the UI can be notified whenever the mode changes: The `VisibleQuestSource` would be one of those listeners, the TBD UI component that shows whether team mode is on (and in what team one is) would be another.

**Comment by HolgerJeromin** [comment]
+1 for colors instead of animals.

**Comment by westnordost** [comment]
If you need more input, or if you are stuck, feel free to press the "request review" button. Currently I have not so much time, so I don't monitor all the tickets.

**Comment by peternewman** [comment]
>     * I am not sure if I like the "animal avatars". The idea to not let users choose "a number from 1 to number of people" is good and fair, but introducing the concept of avatars may add some confusion as avatars come with certain expectations tied to them. It may be more clear to let the user **choose a color**: red, blue, green, yellow, black, orange, violet etc.

Colour-blindness and how different phones render different colours (and dark mode)?

How about shapes or patterns instead?

**Comment by FloEdelmann** [comment]
Thanks @westnordost.

I think I'm done with the UI and saving the preference, but the actual quest filtering is still missing. In [`VisibleQuestsSource`](https://github.com/streetcomplete/StreetComplete/pull/2557/commits/fd71e5fdc8d2813b648ce13ecaef5880488591a9) I have either access to whether team mode is enabled or not, or to which quests are visible. I don't know how/where to combine that information. A little help here would be much appreciated :blush:

@peternewman I also though about color blindness, so I added letters to the color circles. Also, the positions of the circles are now consistent (it doesn't make sense to randomize these anymore like I had done with the animal avatars), so there are three factors to identify (and match) an option: color, letter and position.

Here are some updated and new screenshots :tada: 

### Team mode dialog

 

### When team mode is activated

**Comment by matkoniecz** [comment]
> Seems like I can't "request a review" from you though, since you have not reviewed the PR (I can only click "re-request review" then).

I could press this button, done

**Comment by matkoniecz** [comment]
Is there something in main view that reminds that team mode is enabled? Reminder that only part of quests is shown would be useful.

**Comment by FloEdelmann** [comment]
> something in main view that reminds that team mode is enabled

The chosen team mode color circle is displayed as a little badge for the main menu button (see the third screenshot in https://github.com/streetcomplete/StreetComplete/pull/2557#issuecomment-778077358). Is that too subtle?

**Comment by matkoniecz** [comment]
I missed it, I though about something like following (feel free to ignore me, I am just trying to simulate what may go wrong with confused/slightly clueless user).

Untitled

**Review by westnordost — CHANGES_REQUESTED** [review]
Regarding the VisibleQuestsSource, locally on my branch it looks completely different now, so I'll just delay advice on this until the other things have been addressed I found in the review.

In short, all `Quest`s that are output from that class must be filtered through the `TeamModeFilter` before they are being returned. Both when quests are being queried by bbox and also in listeners.

**Comment by westnordost** [comment]
Oh yea, additionally I concur with @matkoniecz that there should be some kind of heads-up that you are (still) in team mode.

I suggest a simple toast each time the main menu button with the little annotation becomes visible: "Team mode active"

**Comment by HolgerJeromin** [comment]
Teams, Teamviewer, Chrome paint a thick border around the screen when there is an active screenshare. Same could be done here:
image

**Comment by matkoniecz** [comment]
I admit that at least for me it would be not be recognizable as indicator what is happening (as person who never used screenshare so far).

File: app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt
Comment (FloEdelmann) [reply app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt:22]: I think it should rather be `teamSize` and `indexInTeam`.

There will only be one team conceptually (the group of people who are mapping the same area together). When they all enable team mode, everyone in the group has a unique index (in the team).

File: app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt
Comment (FloEdelmann) [reply app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt:19]: It did work in the emulator, but I'm fine with changing it.

File: app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt:22]: Sure, whatever you think is not ambiguous. "Group" and "Team" are interchangable terms then?

File: app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt:19]: Hmm... weird .... who injects the prefs then? Then this is an automatism I didn't know about. Anyway, that lateinit var stuff is just a workaround for Android doing the instantiation of activities and fragments and thus normal dependency injection via the constructor is not possible.

**Comment by peternewman** [comment]
> I admit that at least for me it would be not be recognizable as indicator what is happening (as person who never used screenshare so far).

Presumably in conjunction with something in the menu it's just another nudge though.

File: app/src/main/res/values/strings.xml
Comment (peternewman) [inline app/src/main/res/values/strings.xml:831]: Sorry but this is rather Germanic, one of these two is probably more English.
```suggestion
    <string name="team_mode_team_size_label">How many people are there?</string>
```
```suggestion
    <string name="team_mode_team_size_label">How many people in your team?</string>
```

File: app/src/main/res/values/strings.xml
Comment (peternewman) [inline app/src/main/res/values/strings.xml:832]:
```suggestion
    <string name="team_mode_team_size_hint">Team mode works for groups of between 2 and 12 people.</string>
```

**Review by peternewman — COMMENTED** [review]
Some language improvements

**Comment by peternewman** [comment]
> @peternewman I also thought about color blindness, so I added letters to the color circles. Also, the positions of the circles are now consistent (it doesn't make sense to randomize these anymore like I had done with the animal avatars), so there are three factors to identify (and match) an option: color, letter and position.

I considered suggesting letters or numbers, but was wary of how it fits with people using non-Latin scripts, but I guess it's just a symbol at worst, and combined with the colour and position should cover it for all users.

Although position becomes irrelevant when selected.

File: app/src/main/res/values/strings.xml
Comment (peternewman) [inline app/src/main/res/values/strings.xml:833]: It's not just a unique colour now, but maybe unique dot is more confusing?

File: app/src/main/res/values/strings.xml
Comment (westnordost) [reply app/src/main/res/values/strings.xml:831]: What about "How many are you?" I mean, what else?

File: app/src/main/res/values/strings.xml
Comment (westnordost) [reply app/src/main/res/values/strings.xml:833]: No I think color is better. You suggested yourself to show a boundary in the color chosen.

Comment (peternewman) [reply app/src/main/res/values/strings.xml:833]: Yeah fair enough.

File: app/src/main/res/values/strings.xml
Comment (peternewman) [reply app/src/main/res/values/strings.xml:831]: It still doesn't sound quite right to a native.

If you want something shorter:
```suggestion
    <string name="team_mode_team_size_label">How many people?</string>
```

Comment (peternewman) [reply app/src/main/res/values/strings.xml:831]: Or
```suggestion
    <string name="team_mode_team_size_label">How many of you are there?</string>
```

Comment (peternewman) [reply app/src/main/res/values/strings.xml:831]: https://ell.stackexchange.com/questions/181748/how-many-are-you-guys-or-how-many-people-are-you-guys-which-one-is-correct

File: app/src/main/res/values/strings.xml
Comment (FloEdelmann) [reply app/src/main/res/values/strings.xml:832]: Does this sound a bit nicer?

```suggestion
    <string name="team_mode_team_size_hint">Team mode works for groups of 2 to 12 people.</string>
```

Comment (peternewman) [reply app/src/main/res/values/strings.xml:832]: It certainly sounds less wordy!

File: app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuDialog.kt
Comment (FloEdelmann) [inline app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuDialog.kt:19]: Either I pass the quest filter in here, so that I can access its `enableTeamMode`/`disableTeamMode` functions and `isEnabled`/`indexInTeam` properties, or I pass them as separate parameters. Do you prefer callbacks and separate parameters here, too?

**Review by FloEdelmann — COMMENTED** [review]
~I think all your comments should be resolved now.~ Oh, I just remembered that something is missing: Using the programmatically created color circles in the main menu button and dialog.

If it makes sense for me to wait until your branch is merged, so that we can avoid merge conflicts in `VisibleQuestsSource`, just let me know.

**Comment by westnordost** [comment]
> If it makes sense for me to wait until your branch is merged, so that we can avoid merge conflicts in VisibleQuestsSource, just let me know.

Nah, my refactor work will probably take longer than this PR. The only thing I am concerned about is that it may not be received favourable to launch a feature that promotes to meet up in groups in mids of the pandemic. The monthly OSM "Stammtische", mapathons and other meet-ups are/should also not happening right now in real life. Surely, such feature could be used by flatmates or couples to have something to do on a walk.

File: app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuDialog.kt
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuDialog.kt:19]: I have no preference here. Well... maybe slightly twards callbacks, because then the control-logic is all in one place (in MainMenuButtonFragment)

**Comment by smichel17** [comment]
> The only thing I am concerned about is that it may not be received favourable to launch a feature that promotes to meet up in groups in midst of the pandemic.

On the other hand, going on a walk outdoors is one of the safest in-person activities you can do with someone outside your bubble during the pandemic. Plenty of ventillation, easy to maintain distance, and wearing a mask in the winter can even be nice, to keep your face warm :) I think it would be okay as long as the changelog included a note acknowledging the pandemic.

An alternative would be to merge the PR but comment out the pieces of the UI, so the feature cannot be accessed. Then you can enable the feature later, but avoid extra work resolving merge conflicts.

File: app/src/main/java/de/westnordost/streetcomplete/data/quest/VisibleQuestsSource.kt
Comment (FloEdelmann) [inline app/src/main/java/de/westnordost/streetcomplete/data/quest/VisibleQuestsSource.kt:76]: Here we have access to whether team mode is enabled or not, but how can we loop through all quests and update their visibility then?

**Review by FloEdelmann — COMMENTED** [review]
I've rebased the PR onto the `master` branch and cleaned up / combined the commits to make them as easy to review as possible.

My question on how to proceed remains: In `VisibleQuestsSource` I have either access to whether team mode is enabled or not, or to which quests are visible. I don't know how/where to combine that information.

File: app/src/main/java/de/westnordost/streetcomplete/data/quest/VisibleQuestsSource.kt
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/data/quest/VisibleQuestsSource.kt:76]: Right, this class doesn't know that. So that is why when the switch is flipped, the class should get all the visible quests from the two controllers by itself, sort them into invisible and visible quests and call the `onUpdatedVisibleQuests` listener with that info.
Yes, this may lead to already visible quests being added twice and already not-visible quests to be deleted from view twice. You need to check if the listener(s), mainly it should be the QuestPinLayerManager, are OK with that.

Hmm, but then of course,... the question is fro which bbox the visible quests should be retrieved from DB. Hmm.

I have had the same problem on my development branch, this is why I extended the listener like this:
```kotlin
    interface Listener {
        /** Called when given quests in the given group have been added/removed */
        fun onUpdatedVisibleQuests(added: Collection<Quest>, removed: Collection<Long>, group: QuestGroup)
        /** Called when something has changed which should trigger any listeners to update all */
        fun onInvalidated()
    }
```
... with an onInvalidated() method. The `QuestPinLayerManager` does this then:

```kotlin
    override fun onInvalidated() {
        clear()
        onNewScreenPosition()
    }
```

to re-fetch the new quests.

File: app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuButtonFragment.kt
Comment (FloEdelmann) [inline app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuButtonFragment.kt:89]: Should we invalidate the quests here when team mode gets disabled, as suggested in the original issue #2189?

If so, is `downloadedTilesDao.removeAll()` the right function to call? It doesn't seem start a new download immediately though.

**Review by FloEdelmann — COMMENTED** [review]
Filtering quests now works :tada:

Thanks @westnordost for your guidance! I've got one more question, otherwise this is ready to review :blush:

File: app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuButtonFragment.kt
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuButtonFragment.kt:89]: No, that's not the right function to call. That makes the app forget that it downloaded anything, so it downloads again everything on the next opportunity.

On invalidation, the `QuestPinLayerManager` which is a listener to the `VisibleQuestsSource` should simply call `clear()` on itself (=removes all the quests + forgets that it retrieved from database any quests). This should next time the map is moved re-load all the quests from db, i.e. make `QuestPinLayerManager` to ask `VisibleQuestsSource` for quests.
Maybe to make this happen immediately and not just after the user panned the map, call `onNewScreenPosition()` after `clear()`.

Comment (FloEdelmann) [reply app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuButtonFragment.kt:89]: > That makes the app forget that it downloaded anything, so it downloads again everything on the next opportunity.

That's the idea though. From #2189:

> After exiting the team modus again, those areas' quest cache should be invalidated automatically to make sure that the quests solved by the other team members won't be displayed the next time the user want to map there.

So we actually want to re-download the quests to make sure that other team members' solved quests don't show up after exiting team mode.

The changes in `QuestPinLayerManager` are already implemented in [`68d3a2c` (#2557)](https://github.com/streetcomplete/StreetComplete/pull/2557/commits/68d3a2cce0bf82ae2fc8a63371d5221acb595279).

Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuButtonFragment.kt:89]: > That's the idea though. From #2189:

>>    After exiting the team modus again, those areas' quest cache should be invalidated automatically to make sure that the quests solved by the other team members won't be displayed the next time the user want to map there.

Ah right, makes sense. Hmm, but it only "invalidates" them, so they won't actually get updated after the next download. The `QuestAutoSyncer ` does the automatic downloading. So, that class should maybe also listen to changes on the team mode state. If the team mode got turned off, it should trigger an auto-download (after `downloadedTilesDao.removeAll()` has been called).

**Comment by Helium314** [comment]
I did a short test of the team mode (team of 2), as it looks like it is almost finished.
Some comments:

1. When enabling team mode, the field for entering team size should have a blinking cursor (and the keyboard could pop up immediately, as it is needed anyway)

2. I don't like the changed position of the _settings_ entry in the menu, it's a bit of an unnecessary change in th UI. But this may just be personal taste, I guess after a few days/weeks I will get used to the layout

3. thank you! It works well!

**Comment by westnordost** [comment]
I noticed you committed something here. I lost track - is it ready to merge? Is another review required?

**Comment by RubenKelevra** [comment]
Not sure what's wrong with the animals as avatars... they looked really cute. Just just A, B, C, D is pretty boring tbh. 

Just my two cents. 

Thanks for the efforts implementing my ticket! 😁

**Comment by westnordost** [comment]
> By the way, do we have a quest counter for the team mode? I feel like it would be cool to be able to compare the progress for each team member (on their phone) instead or showing the global counter. :)

If you add something like this later to a post, it will be missed by people who only read the updates via email. Anyway, as far as I know, Flo didn't implement anything like that. I agree that a feature like that might be in the spirit of that mode. But something like this could always be added later.

**Comment by RubenKelevra** [comment]
> > By the way, do we have a quest counter for the team mode? I feel like it would be cool to be able to compare the progress for each team member (on their phone) instead or showing the global counter. :)
> 
> If you add something like this later to a post, it will be missed by people who only read the updates via email. Anyway, as far as I know, Flo didn't implement anything like that. I agree that a feature like that might be in the spirit of that mode. But something like this could always be added later.

I just tapped the "comment" button accidentally and didn't thought about the mail users. Next time I'll just create the full comment. 😁

But I agree that this is a nice-to-have and shouldn't stop this PR. 


@FloEdelmann I've posted the first pictures posted here as an appetizers on Twitter without realizing that the approach changed and those animals won't be used. My bad - sorry!

Anyway: There was a pretty good response to that news!

How about creating a poll there if they prefer the Animals or the Alphabetical Avatars? 🤔

**Comment by westnordost** [comment]
Nah, no poll. The reasons for choosing colors over animals are:
- slightly better distinguishable. There are many brownish animal heads, some of these could not be told apart on first sight
- better communicable. "I take blue" is easer than "I take the light brown animal that looks like there are little crosses on his ears"
- better recognizable when displayed small (i.e. above the main menu button)
- consistent with other games: in boardgames, and computer games, the team is usually always defined by a color; which character/faction you choose is a different matter and has usually nothing to do with in which team you are but has other perks. This is why I mentioned that choosing avatars comes with other expectations connected with it that may lead to confusion
- you already got an OSM avatar. Confusion might arise in what way these two interact, if at all. Of course they have nothing to do with each other, but it may still lead to confusion. Users are easily confused.

The letters on the color was @FloEdelmann 's decision, I guess he added it so it is even more easy to communicate in which team you are. So instead of saying "I take aquamarine", you can say "I take the D" or something like that. I am neutral towards that and slightly lean towards not even having letters there, but just colors (that then must be very easy to tell apart). But this can all be changed later. In case we want to keep the labels, I'd also slightly lean towards numbers instead of letters because numbers are more international than latin letters, I'd even go so far that these are understood and used everywhere, even if the script used in that place has own numbers (such as in Arabia, India and South East Asia).

**Comment by matkoniecz** [comment]
> The letters on the color was @FloEdelmann 's decision, I guess he added it so it is even more easy to communicate in which team you are. So instead of saying "I take aquamarine", you can say "I take the D" or something like that. I am neutral towards that and slightly lean towards not even having letters there, but just colors (that then must be very easy to tell apart).

letters may be useful for colorblind people (also, in some board games there are also shapes in addition to colors - see https://www.dangergames.co/blog/designing-games-for-colorblind-players/ )

screen06

**Comment by peternewman** [comment]
> > The letters on the color was @FloEdelmann 's decision, I guess he added it so it is even more easy to communicate in which team you are. So instead of saying "I take aquamarine", you can say "I take the D" or something like that. I am neutral towards that and slightly lean towards not even having letters there, but just colors (that then must be very easy to tell apart).
> 
> letters may be useful for colorblind people (also, in some board games there are also shapes in addition to colors - see https://www.dangergames.co/blog/designing-games-for-colorblind-players/ )

Yeah we shouldn't be adding features which actively inhibit colour-blind users in this day and age! They will also deal with people's screens having different rendering/brightness/contrast especially in bright light, to ensure you can tell the potential 12 colours apart. But as you say letters, numbers or even basic symbols (e.g. square, triangle, circle, star etc) would do the job. I wonder if a screen reader would cope with the letters or numbers as is...

Also in terms of using just colours, see https://en.wikipedia.org/wiki/The_dress and https://colornames.org/ and similar, one person's aquamarine is another person's blue!

**Comment by FloEdelmann** [comment]
Thanks for all your comments!

@westnordost Yes, this is ready for review. I'll try to test on real devices in the next few days.

@RubenKelevra Regarding the animal avatars: I also thought they were a cute idea to make the selection more playful. However, additional to the problems that @westnordost noted: I'd like to have a bit of variation between avatars, which is not that easy. This variation can be based the team size (but then you'd still get the same two avatars every time you're mapping with two people) and/or the time. You can see how I implemented that [here](https://github.com/streetcomplete/StreetComplete/commit/3cd9cb18cc0ac150529ee56786f721649df8238f#diff-cf4dee20bc508dfbf2d951317003aec9d6b303d34e8bad457b03c825ea3dc266R101-R107) (the `total` parameter is the team size there). Being based on time, it inherently has the problem of syncing: Two phones may have slightly different times set; or one person just happens to open the dialog right before a new selection is chosen, and the other one right after that. Then they end up with different sets of avatars.

Making the selection more boring with color circles also avoids all of these problems.

@RubenKelevra Regarding the team mode counter: I won't implement this in this PR. Let's first find out together if this works as expected; then it can later be improved. Good idea though!

Regarding the letters on the color circles: As @peternewman already said: There needs to be an additional way to distinguish the color circles for colorblind people, but also to mitigate differences in displayed colors. I don't really care whether it's latin letters, arabic digits, or other symbols. Coming up with 12 distinct shapes that still are simple and easy to recognize is actually quite hard though. Also, using letters or digits is much easier to code and scale (e.g. we might decide to increase the maximum team size to 15, then we'd need to come up with 3 more shapes).

**Comment by RubenKelevra** [comment]
@FloEdelmann just use the day of the month for the avatar variation. It's unlikely that two people have two different dates in their phone. :)

**Comment by peternewman** [comment]
> @FloEdelmann just use the day of the month for the avatar variation. It's unlikely that two people have two different dates in their phone. :)

Timezones and people like Australia/NZ may be a day ahead depending on how they're running their phone with local timezone or world clock and a location!

> @RubenKelevra Regarding the animal avatars: I also thought they were a cute idea to make the selection more playful. However, additional to the problems that @westnordost noted: I'd like to have a bit of variation between avatars, which is not that easy. This variation can be based the team size (but then you'd still get the same two avatars every time you're mapping with two people) and/or the time.

I would have thought the variation might be a disadvantage personally. E.g. when playing a colour based board game, or a counter based one (e.g. Monopoly), in a regular group of players certain people are always likely to take certain choices so it almost becomes a given, which removes choosing time and also means you don't get confused when who you are (although obviously the latter bit isn't relevant to SC once you've made your choice). If that changes each time you've got to pick and communicate that pick each time.

**Comment by westnordost** [comment]
@FloEdelmann: I am currently reviewing it and making small changes to it. If you don't mind, I'd also change from letters to numbers because I think it is slightly better. Ok?

**Comment by westnordost** [comment]
Oh, I get now why you chose letters over numbers. A number is confusing when shown as a little colored circle on top of the main menu button

**Review by westnordost — APPROVED** [review]

**Comment by RubenKelevra** [comment]
> > @FloEdelmann just use the day of the month for the avatar variation. It's unlikely that two people have two different dates in their phone. :)
> 
> Timezones and people like Australia/NZ may be a day ahead depending on how they're running their phone with local timezone or world clock and a location!


Just ignore the timezones. UTC should be everywhere the same.

If this isn't fault proof enough I would recommend asking the OSM API for the current time. I think HTTP is still posting the current time on each response, no? 🤔

**Comment by peternewman** [comment]
> Just ignore the timezones. UTC should be everywhere the same.

Yeah fair point, although doesn't deal with the IMHO other downsides.

> If this isn't fault proof enough I would recommend asking the OSM API for the current time. I think HTTP is still posting the current time on each response, no? thinking

That doesn't cover the users who use it offline and just sync via Wi-Fi occasionally/when they return home, which is effectively something the whole of the rest of this PR enables, otherwise you could just keep uploading and rescanning I think.

File: app/src/main/res/values/strings.xml
Comment (FloEdelmann) [reply app/src/main/res/values/strings.xml:831]: @westnordost Why did you change this back to the string I had initially?

File: app/src/main/java/de/westnordost/streetcomplete/controls/TeamModeColorCircleView.kt
Comment (FloEdelmann) [inline app/src/main/java/de/westnordost/streetcomplete/controls/TeamModeColorCircleView.kt:17]: Shouldn't this be `setIndexInTeam` again to avoid confusion?

File: app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt
Comment (FloEdelmann) [inline app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt:64]: I thought you didn't want to hardcode this but return the colors array size instead?

**Review by FloEdelmann — COMMENTED** [review]
Thanks @westnordost for your changes! I got a few post-merge questions though.

File: app/src/main/java/de/westnordost/streetcomplete/controls/TeamModeColorCircleView.kt
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/controls/TeamModeColorCircleView.kt:17]: Hm, right

File: app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt:64]: Well, now that the Colors object is gone (converted into view), it seemed wrong to me to have this information sourced from a view class. So I thought I put it where the TeamMode logic has its core.

File: app/src/main/res/values/strings.xml
Comment (westnordost) [reply app/src/main/res/values/strings.xml:831]: Oh, it's just that it appeared to be better in my ears, I didn't remember this discussion

**Comment by RubenKelevra** [comment]
> > Just ignore the timezones. UTC should be everywhere the same.
> 
> Yeah fair point, although doesn't deal with the IMHO other downsides.
> 
> > If this isn't fault proof enough I would recommend asking the OSM API for the current time. I think HTTP is still posting the current time on each response, no? thinking
> 
> That doesn't cover the users who use it offline and just sync via Wi-Fi occasionally/when they return home, which is effectively something the whole of the rest of this PR enables, otherwise you could just keep uploading and rescanning I think.

I just want the cute Animals back 😬

I mean we don't change the colors either. So two people has to be red and blue too all the time, right? 🤔

The A, B, C, D is just outright boring.

If you think daily changes are bad, just do it weekly. Hard to belive that people go team mapping on Sunday nights with phones having clocks completely off. 😬

File: app/src/main/res/values/strings.xml
Comment (FloEdelmann) [reply app/src/main/res/values/strings.xml:831]: Fixed in #2613.

File: app/src/main/java/de/westnordost/streetcomplete/controls/TeamModeColorCircleView.kt
Comment (FloEdelmann) [reply app/src/main/java/de/westnordost/streetcomplete/controls/TeamModeColorCircleView.kt:17]: Fixed in #2613.

File: app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt
Comment (FloEdelmann) [reply app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/TeamModeQuestFilter.kt:64]: Fixed in #2613.

**Pull request merged**
